go/vt/vttablet/tabletmanager/vreplication/replicator_plan.go

```diff
@@ -444,6 +475,61 @@ func (tp *TablePlan) applyChange(rowChange *binlogdatapb.RowChange, executor fun
 		if tp.isOutsidePKRange(bindvars, before, after, "insert") {
 			return nil, nil
 		}
+		if tp.isPartial(rowChange) {
+			// We need to use a combination of the values in the BEFORE and AFTER image to generate the
+			// new row.
+			jsonIndex := 0
+			for i, field := range tp.Fields {
+				if field.Type == querypb.Type_JSON && rowChange.JsonPartialValues != nil {
+					switch {
+					case !isBitSet(rowChange.JsonPartialValues.Cols, jsonIndex):
+						// We use the full AFTER value which we already have.
+					case len(afterVals[i].Raw()) == 0:
+						// If the JSON column was NOT updated then the JSON column is marked as partial
+						// and the diff is empty as a way to exclude it from the AFTER image. So we
+						// want to use the BEFORE image value.
+						beforeVal, err := vjson.MarshalSQLValue(bindvars["b_"+field.Name].Value)
+						if err != nil {
+							return nil, vterrors.Wrapf(err, "failed to convert JSON to SQL field value for %s.%s when building insert query",
+								tp.TargetName, field.Name)
+						}
+						bindvars["a_"+field.Name], err = tp.bindFieldVal(field, beforeVal)
+						if err != nil {
+							return nil, vterrors.Wrapf(err, "failed to bind field value for %s.%s when building insert query",
+								tp.TargetName, field.Name)
+						}
+					default:
+						// For JSON columns when binlog-row-value-options=PARTIAL_JSON is used and the
+						// column is marked as partial, we need to wrap the JSON diff function(s)
+						// around the BEFORE value.
+						diff := afterVals[i].RawStr()
+						beforeVal := bindvars["b_"+field.Name].Value
+						buf := bytes.Buffer{}
+						buf.Grow(len(diff) + len(beforeVal) + len(sqlparser.Utf8mb4Str) + 2) // +2 is for the enclosing quotes
+						buf.WriteString(sqlparser.Utf8mb4Str)
+						buf.WriteByte('\'')
+						buf.Write(beforeVal)
+						buf.WriteByte('\'')
+						newVal := sqltypes.MakeTrusted(querypb.Type_EXPRESSION, []byte(
+							fmt.Sprintf(diff, buf.String()),
+						))
+						bv, err := tp.bindFieldVal(field, &newVal)
+						if err != nil {
+							return nil, vterrors.Wrapf(err, "failed to bind field value for %s.%s when building insert query",
+								tp.TargetName, field.Name)
+						}
+						bindvars["a_"+field.Name] = bv
+					}
+					jsonIndex++
+					continue
+				}
+				if !isBitSet(rowChange.DataColumns.Cols, i) {
+					return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL,
+						"binary log event missing a needed value for %s.%s due to not using binlog-row-image=FULL; you will need to re-run the workflow with binlog-row-image=FULL",
+						tp.TargetName, field.Name)
+				}
```

Comment:
This fixes a bug we had with NOBLOB support. Without this, the blob/text field value would be lost.